### PR TITLE
Fix network connection state tracking

### PR DESCRIPTION
**Pull request by MatkovIvan**
* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Fix network connection state tracking

## Related PRs or issues

#941

### CODE REVIEW

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java [hunk 1]
```diff
@@ -149,81 +144,54 @@ private IntentFilter getOldIntentFilter() {
...
-    /**
-     * Handle network available update on API level >= 21.
-     */
-    private synchronized void onNetworkAvailable(Network network) {
-        AppCenterLog.debug(LOG_TAG, "Network available netId: " + network);
-        mAvailableNetworks.add(network);
-        AppCenterLog.debug(LOG_TAG, "Available networks netIds: " + mAvailableNetworks);
+        /* The actual current state should be checked here instead of using a value based on connection change events. */
+        try {
+            NetworkInfo info = mConnectivityManager.getActiveNetworkInfo();
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: No this is inaccurate and caused issues in the past, we had to move away from this call. Please revert this change.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java [hunk 2]
```diff
@@ -42,6 +42,9 @@ public synchronized ServiceCall callAsync(String url, String method, Map<String,
         Call call = new Call(mDecoratedApi, url, method, headers, callTemplate, serviceCallback);
         if (mNetworkStateHelper.isNetworkConnected()) {
             call.run();
+
+            /* Send pending call(s), if any. It can be the case if a network state update event is missing. */
+            onNetworkStateUpdated(true);
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java]: This is not the right solution I think.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java (hunk 1 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: Maybe keep it as a fallback? Because rely just on connection change events is inaccurate too

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java (hunk 2 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java]: I agree that we can play with network capabilities, but this might be saved as a fallback.

Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java]: But isNetworkConnected returns false if there is no event with the modern APIs, also the old APIs might work fine on older devices, the customer is on API level 23 apparently so using the new APIs.

Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java]: (also getActiveNetworkInfo is out of sync with events on some Android versions and has been deprecated, specifically on Android 8 where it can say "blocked" while we received an event for active using old APIs)

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java (hunk 1 above)
Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: This API has already been proven to not work correctly like returning "blocked" state while the network was up. It is also deprecated. It also does not work correctly when there is more than 1 network as mentioned in the past by a customer who worked on a network library.

Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: it's not deprecated, but OK, I got it. Thanks for the info

**Comment by guperrot** [comment]
We can consider the fallbacks later if nothing else seems to work (after investigating the capability filters). But in that case we would better poll all network info and not the "active" one.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java (hunk 2 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java]: it's not deprecated. I tested on Android 9 and yes, it's out of sync with events, but it's not an issue here.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java (hunk 1 above)
Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: Ah yeah its `getAllNetworkInfo()` that is deprecated indeed.

**Comment by MatkovIvan** [comment]
List of changes:
* Do not call network connected state change event on change type/number of connected networks (there was no action on it anyway);
* Remove `NetworkCapabilities.NET_CAPABILITY_VALIDATED` to get connection events to restricted networks (the main fix);
* Check all network state as fallback to make sure if we are really offline;
* Send queued logs even if connection event is missed;
* Threat lack of network state permission as "always online";
* Remove logging network ids (can be treated as sensitive information);

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java (hunk 2 above)
Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/http/HttpClientNetworkStateHandler.java]: I'm still uncomfortable with this line. It's code executed from a specific consumer of the network state helper lib and was not needed to fix the issue. Can we revert that line?

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java [hunk 3]
```diff
@@ -123,17 +121,20 @@ public void onLost(Network network) {
...
 
             /*
              * Can be security exception if permission missing or sometimes another runtime exception
              * on some customized firmwares.
              */
-            AppCenterLog.error(LOG_TAG, "Cannot access network state information", e);
+            AppCenterLog.error(LOG_TAG, "Cannot access network state information.", e);
+
+            /* We should try to send the data, even if we can't get the current network state. */
+            mConnected = true;
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: ok

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java [hunk 4]
```diff
@@ -149,81 +150,73 @@ private IntentFilter getOldIntentFilter() {
      * @return true for connected, false for disconnected.
      */
     public synchronized boolean isNetworkConnected() {
-        return mNetworkType != null || !mAvailableNetworks.isEmpty();
+        return mConnected || isAnyNetworkConnected();
     }
 
     /**
-     * Handle network available update on API level >= 21.
+     * Check if any network is connected.
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: Maybe comment it's used only as a fallback.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java [hunk 5]
```diff
@@ -149,81 +150,73 @@ private IntentFilter getOldIntentFilter() {
...
+    private synchronized void onNetworkAvailable(@SuppressWarnings("unused") Network network) {
+        if (!mConnected) {
             notifyNetworkStateUpdated(true);
+            mConnected = true;
         }
     }
 
     /**
-     * Update network type by polling on API level < 21.
+     * Handle network available update on API level >= 21.
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: network lost update, not available

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java [hunk 6]
```diff
@@ -149,81 +150,73 @@ private IntentFilter getOldIntentFilter() {
...
-
-        /* Update network type. null for not connected. */
-        if (networkInfo != null && networkInfo.isConnected()) {
-            mNetworkType = networkInfo.getTypeName() + networkInfo.getSubtypeName();
-        } else {
-            mNetworkType = null;
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private synchronized void onNetworkLost(Network network) {
+        Network[] networks = mConnectivityManager.getAllNetworks();
+        boolean noNetworks = networks == null || networks.length == 0 ||
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: remove `s` from `noNetworks`

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java [hunk 7]
```diff
@@ -149,81 +150,73 @@ private IntentFilter getOldIntentFilter() {
...
+                Arrays.equals(networks, new Network[] { network });
+        if (mConnected && noNetworks) {
+            notifyNetworkStateUpdated(false);
+            mConnected = false;
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: update state before notifying

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java [hunk 8]
```diff
@@ -149,81 +150,73 @@ private IntentFilter getOldIntentFilter() {
...
-        boolean networkTypeChanged = previousNetworkType == null ? mNetworkType != null : !previousNetworkType.equals(mNetworkType);
-        if (networkTypeChanged) {
-            boolean connected = isNetworkConnected();
-            if (connected && previousNetworkType != null) {
-                notifyNetworkStateUpdated(false);
-            }
+        boolean connected = isAnyNetworkConnected();
+        if (connected != mConnected) {
             notifyNetworkStateUpdated(connected);
+            mConnected = connected;
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: update state before notifying

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java (hunk 4 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: https://github.com/Microsoft/AppCenter-SDK-Android/pull/943/files#diff-f8b58324d3d38f4997ba0df305858c37R216

**Comment by guperrot** [comment]
What API levels were tested on real device?

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java [hunk 9]
```diff
@@ -233,19 +226,19 @@ private synchronized void handleNetworkStateUpdate() {
      * @param connected whether the network is connected or not.
      */
     private void notifyNetworkStateUpdated(boolean connected) {
+        AppCenterLog.debug(LOG_TAG, "Network has been " + (connected ? "connected." : "disconnected."));
```
Comment (MatkovIvan) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: @guperrot It's logged here. Anyway, I'll revert the logs in the callbacks

Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: I couldn't see any debug logs when I was testing, will retry.

**Comment by guperrot** [comment]
Looks good, need more testing now.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java (hunk 9 above)
Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/NetworkStateHelper.java]: I see this log now, maybe I missed it last time.
